File: examples/c/hello_shareware.c
/*
 * hello-shareware — a complete, minimal shareware app in C.
 *
 * It looks for a license key on disk, verifies it offline against the omarket
 * platform public key, and unlocks a feature when the key is genuine and
 * belongs to this app. No account, no activation call, no phone-home:
 * everything below runs with the network unplugged.
 *
 * A SHRW1 key is three dot-separated fields:
 *
 *     SHRW1.<base64url(payload JSON)>.<base64url(ed25519 signature)>
 *
 * Both segments are base64url with the padding stripped. The signature covers
 * the *exact payload bytes* — decode segment 2, verify over those bytes, and
 * only then parse the JSON. Parsing first and verifying over a re-encoding of
 * the result is the classic way to break this: key order and spacing would
 * have to match byte for byte, and one day they will not.
 *
 *   Build:  cc -O2 -Wall -Wextra -o hello-shareware hello_shareware.c -lcrypto
 *   Needs:  OpenSSL 1.1.1 or newer (Ed25519 landed there).
 */

#define _POSIX_C_SOURCE 200809L

#include <openssl/evp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* APP_ID must match the "app" field inside the license payload. This is what
   stops a valid key for someone else's app from unlocking yours. */
#define APP_ID "hello-shareware"
#define APP_NAME "Hello Shareware"

/* The omarket Ed25519 license-signing key, also served at
   https://omarket.dev/api/pubkey. Bake it into your binary — that is what
   makes verification offline. It is a public key: shipping it to users is the
   point, and it cannot be used to mint licenses. */
static const char PLATFORM_PUBLIC_KEY[] = "vIODssCr6I8zWUTaK/IhdHzkoK5LTdEzbtIeXRnatko=";

#define ED25519_PUBLIC_KEY_LEN 32
#define ED25519_SIGNATURE_LEN 64
#define KEY_MAX 4096
#define PAYLOAD_MAX 2048

struct license {
	char id[64];
	char app[96];
	char kind[32];
	long long issued_at;
};

/* ---------------------------------------------------------------- base64 */

/* b64_value maps one base64 character to its 6 bits, accepting both the
   standard (+/) and URL-safe (-_) alphabets — so one decoder handles the
   padded standard base64 of the public key and the unpadded base64url of the
   two key segments. Returns -1 for anything else. */
static int b64_value(unsigned char c)
{
	if (c >= 'A' && c <= 'Z') return c - 'A';
	if (c >= 'a' && c <= 'z') return c - 'a' + 26;
	if (c >= '0' && c <= '9') return c - '0' + 52;
	if (c == '+' || c == '-') return 62;
	if (c == '/' || c == '_') return 63;
	return -1;
}

/* b64_decode decodes len characters from in into out, writing at most cap
   bytes and storing the count in *out_len. Padding is optional. Returns 1 on
   success, 0 on an unexpected character or on overflow. */
static int b64_decode(const char *in, size_t len, unsigned char *out, size_t cap,
                      size_t *out_len)
{
	unsigned int acc = 0;
	int bits = 0;
	size_t n = 0;

	for (size_t i = 0; i < len; i++) {
		if (in[i] == '=') break; /* padding; nothing meaningful follows */
		int v = b64_value((unsigned char)in[i]);
		if (v < 0) return 0;
		acc = (acc << 6) | (unsigned int)v;
		bits += 6;
		if (bits >= 8) {
			bits -= 8;
			if (n >= cap) return 0;
			out[n++] = (unsigned char)((acc >> bits) & 0xFF);
		}
	}
	*out_len = n;
	return 1;
}

/* --------------------------------------------------------------- ed25519 */

/* ed25519_verify checks sig (64 bytes) over msg using the raw 32-byte public
   key pub. Ed25519 in OpenSSL is one-shot: EVP_DigestVerify, never the
   Update/Final pair. */
static int ed25519_verify(const unsigned char *pub, const unsigned char *msg,
                          size_t msg_len, const unsigned char *sig)
{
	EVP_PKEY *pkey = EVP_PKEY_new_raw_public_key(EVP_PKEY_ED25519, NULL, pub,
	                                             ED25519_PUBLIC_KEY_LEN);
	if (!pkey) return 0;

	EVP_MD_CTX *ctx = EVP_MD_CTX_new();
	int ok = 0;
	if (ctx && EVP_DigestVerifyInit(ctx, NULL, NULL, NULL, pkey) == 1)
		ok = EVP_DigestVerify(ctx, sig, ED25519_SIGNATURE_LEN, msg, msg_len) == 1;

	EVP_MD_CTX_free(ctx);
	EVP_PKEY_free(pkey);
	return ok;
}

/* ------------------------------------------------------------------ json */

/* The SHRW1 payload is a flat JSON object of short scalars — no nesting, no
   escapes, no arrays — so searching for the quoted field name is enough, and
   keeps this file dependency-free. Do not reach for these three in a program
   that parses JSON whose signature it has not already checked. */

static const char *json_field(const char *p, size_t n, const char *name)
{
	char pat[64];
	int len = snprintf(pat, sizeof pat, "\"%s\":", name);
	if (len <= 0 || (size_t)len >= sizeof pat) return NULL;

	for (size_t i = 0; i + (size_t)len <= n; i++)
		if (memcmp(p + i, pat, (size_t)len) == 0) {
			const char *v = p + i + len;
			while (v < p + n && (*v == ' ' || *v == '\t')) v++;
			return v;
		}
	return NULL;
}

static int json_string(const char *p, size_t n, const char *name, char *out, size_t cap)
{
	const char *v = json_field(p, n, name);
	if (!v || v >= p + n || *v != '"') return 0;

	size_t i = 0;
	for (v++; v < p + n && *v != '"'; v++) {
		if (i + 1 >= cap) return 0;
		out[i++] = *v;
	}
	if (v >= p + n) return 0;
	out[i] = '\0';
	return 1;
}

static int json_number(const char *p, size_t n, const char *name, long long *out)
{
	const char *v = json_field(p, n, name);
	if (!v) return 0;

	char buf[32];
	size_t i = 0;
	for (; v < p + n && (*v == '-' || (*v >= '0' && *v <= '9')); v++) {
		if (i + 1 >= sizeof buf) return 0;
		buf[i++] = *v;
	}
	if (i == 0) return 0;
	buf[i] = '\0';
	*out = strtoll(buf, NULL, 10);
	return 1;
}

/* --------------------------------------------------------------- license */

/* verify_license splits key, checks its signature against the base64 public
   key pub_b64, and fills *out. Returns NULL on success, or a short reason.
   It answers exactly one question — did the platform sign this? — and says
   nothing about which app the key is for. That check belongs to the caller. */
static const char *verify_license(const char *key, const char *pub_b64,
                                  struct license *out)
{
	/* 1. split SHRW1.<payload>.<signature> */
	const char *dot1 = strchr(key, '.');
	if (!dot1 || (size_t)(dot1 - key) != 5 || memcmp(key, "SHRW1", 5) != 0)
		return "not a SHRW1 key";
	const char *dot2 = strchr(dot1 + 1, '.');
	if (!dot2 || strchr(dot2 + 1, '.'))
		return "not a SHRW1 key";

	/* 2. decode the three binary pieces */
	unsigned char payload[PAYLOAD_MAX];
	unsigned char sig[ED25519_SIGNATURE_LEN + 1];
	unsigned char pub[ED25519_PUBLIC_KEY_LEN + 1];
	size_t payload_len, sig_len, pub_len;

	if (!b64_decode(dot1 + 1, (size_t)(dot2 - dot1 - 1), payload, sizeof payload,
	                &payload_len))
		return "malformed payload";
	if (!b64_decode(dot2 + 1, strlen(dot2 + 1), sig, sizeof sig, &sig_len) ||
	    sig_len != ED25519_SIGNATURE_LEN)
		return "malformed signature";
	if (!b64_decode(pub_b64, strlen(pub_b64), pub, sizeof pub, &pub_len) ||
	    pub_len != ED25519_PUBLIC_KEY_LEN)
		return "malformed public key";

	/* 3. the signature covers the payload bytes exactly as they arrived */
	if (!ed25519_verify(pub, payload, payload_len, sig))
		return "bad signature";

	/* 4. only now is the payload worth reading */
	const char *json = (const char *)payload;
	long long version = 0;
	if (!json_number(json, payload_len, "v", &version) || version != 1)
		return "unsupported license version";
	if (!json_string(json, payload_len, "id", out->id, sizeof out->id) ||
	    !json_string(json, payload_len, "app", out->app, sizeof out->app) ||
	    !json_string(json, payload_len, "kind", out->kind, sizeof out->kind) ||
	    !json_number(json, payload_len, "issued_at", &out->issued_at))
		return "incomplete payload";

	return NULL;
}

/* ------------------------------------------------------------------ main */

/* default_license_path builds $XDG_CONFIG_HOME/shareware/licenses/<app>.key,
   falling back to ~/.config — where `omarket buy` writes the key. */
static int default_license_path(char *out, size_t cap)
{
	const char *xdg = getenv("XDG_CONFIG_HOME");
	const char *home = getenv("HOME");
	int n;

	if (xdg && *xdg)
		n = snprintf(out, cap, "%s/shareware/licenses/%s.key", xdg, APP_ID);
	else if (home && *home)
		n = snprintf(out, cap, "%s/.config/shareware/licenses/%s.key", home, APP_ID);
	else
		return 0;

	return n > 0 && (size_t)n < cap;
}

/* read_key_file reads path into out and trims surrounding whitespace — the
   stored key file ends with a newline. */
static int read_key_file(const char *path, char *out, size_t cap)
{
	FILE *f = fopen(path, "rb");
	if (!f) return 0;

	size_t n = fread(out, 1, cap - 1, f);
	fclose(f);
	out[n] = '\0';

	while (n > 0 && (unsigned char)out[n - 1] <= ' ') out[--n] = '\0';
	size_t lead = 0;
	while (out[lead] && (unsigned char)out[lead] <= ' ') lead++;
	if (lead) memmove(out, out + lead, n - lead + 1);
	return 1;
}

static void nag(const char *reason)
{
	printf("  [ ] unregistered — %s\n", reason);
	printf("      buy a key:  omarket buy %s\n\n", APP_ID);
	printf("  the deluxe feature is off until this app is registered.\n");
}

int main(int argc, char **argv)
{
	printf("%s 1.0\n\n", APP_NAME);

	char path[1024];
	if (argc > 1)
		snprintf(path, sizeof path, "%s", argv[1]);
	else if (!default_license_path(path, sizeof path))
		snprintf(path, sizeof path, "%s.key", APP_ID);

	/* SHAREWARE_PUBLIC_KEY overrides the baked-in platform key — how you test
	   against a local stack, or against examples/testdata. */
	const char *pub_b64 = getenv("SHAREWARE_PUBLIC_KEY");
	if (!pub_b64 || !*pub_b64) pub_b64 = PLATFORM_PUBLIC_KEY;

	char key[KEY_MAX];
	char why[1200];

	if (!read_key_file(path, key, sizeof key)) {
		snprintf(why, sizeof why, "no license file at %s", path);
		nag(why);
		return 0;
	}

	struct license lic;
	const char *err = verify_license(key, pub_b64, &lic);
	if (err) {
		nag(err);
		return 0;
	}

	/* Genuine — but is it ours? Skip this and any paid omarket license, for
	   any app, unlocks yours. */
	if (strcmp(lic.app, APP_ID) != 0) {
		snprintf(why, sizeof why, "key is for \"%s\", not \"%s\"", lic.app, APP_ID);
		nag(why);
		return 0;
	}

	char issued[16] = "?";
	time_t t = (time_t)lic.issued_at;
	struct tm tm;
	if (gmtime_r(&t, &tm)) strftime(issued, sizeof issued, "%Y-%m-%d", &tm);

	printf("  [x] registered\n");
	printf("      license  %s\n", lic.id);
	printf("      kind     %s\n", lic.kind);
	printf("      issued   %s\n\n", issued);
	printf("  the deluxe feature is unlocked. Enjoy.\n");
	return 0;
}

File: examples/ruby/hello_shareware.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# hello-shareware — a complete, minimal shareware app in Ruby.
#
# It looks for a license key on disk, verifies it offline against the omarket
# platform public key, and unlocks a feature when the key is genuine and
# belongs to this app. No account, no activation call, no phone-home:
# everything below runs with the network unplugged.
#
# openssl and json are still default gems. base64 is not, as of Ruby 3.4 —
# the sibling Gemfile pins it. Ruby 2.6 or newer.
#
#   bundle install
#   ruby hello_shareware.rb
#   ruby hello_shareware.rb ../testdata/hello-shareware.key

begin
  require "bundler/setup" if File.file?(File.expand_path("Gemfile", __dir__))
rescue LoadError
end

begin
  require "base64"
rescue LoadError
  abort "hello_shareware.rb needs the base64 gem. From examples/ruby: bundle install"
end
require "json"
require "openssl"

# APP_ID must match the "app" field inside the license payload. This is what
# stops a valid key for someone else's app from unlocking yours.
APP_ID = "hello-shareware"
APP_NAME = "Hello Shareware"

# The omarket Ed25519 license-signing key, also served at
# https://omarket.dev/api/pubkey. Bake it into your app — that is what makes
# verification offline. It is a public key: shipping it to users is the point,
# and it cannot be used to mint licenses.
PLATFORM_PUBLIC_KEY = "vIODssCr6I8zWUTaK/IhdHzkoK5LTdEzbtIeXRnatko="

# Raised for every "this key does not unlock the app" outcome. A real app may
# want to tell a malformed key apart from a forged one; both mean unregistered.
class LicenseError < StandardError; end

# The SubjectPublicKeyInfo DER header for an Ed25519 key. OpenSSL::PKey.read
# wants a structured key, not 32 loose bytes, and the wrapper is constant:
#
#   SEQUENCE { SEQUENCE { OID 1.3.101.112 }, BIT STRING { 0x00 || key } }
#
# Ruby 3.4+ ships OpenSSL::PKey.new_raw_public_key, which does this for you.
# Twelve fixed bytes cost less than a minimum-version requirement.
ED25519_SPKI_PREFIX = [0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70,
                       0x03, 0x21, 0x00].pack("C*").freeze

def ed25519_public_key(raw32)
  OpenSSL::PKey.read(ED25519_SPKI_PREFIX + raw32)
end

# decode_b64url decodes an unpadded base64url segment, reporting which one
# went wrong rather than letting an ArgumentError escape.
def decode_b64url(str, what)
  Base64.urlsafe_decode64(str)
rescue ArgumentError
  raise LicenseError, "malformed #{what}"
end

# verify_license splits key, checks its Ed25519 signature against the base64
# public key pub_b64, and returns the decoded payload as a Hash.
#
# It answers exactly one question — did the platform sign this? — and says
# nothing about which app the key is for. That check belongs to the caller.
def verify_license(key, pub_b64)
  # 1. SHRW1.<base64url(payload JSON)>.<base64url(ed25519 signature)>, both
  #    segments unpadded.
  parts = key.split(".")
  raise LicenseError, "not a SHRW1 key" unless parts.length == 3 && parts[0] == "SHRW1"

  _, payload_b64, sig_b64 = parts

  payload = decode_b64url(payload_b64, "payload")
  signature = decode_b64url(sig_b64, "signature")
  raise LicenseError, "malformed signature" unless signature.bytesize == 64

  # 2. The signature covers the *exact payload bytes*. Verify over `payload`
  #    as decoded — never over JSON.generate(JSON.parse(payload)), which would
  #    have to reproduce key order and spacing byte for byte, and one day
  #    would not.
  pub_raw = Base64.strict_decode64(pub_b64)
  raise LicenseError, "malformed public key" unless pub_raw.bytesize == 32

  # `verify(nil, ...)` — Ed25519 signs the message itself, with no separate
  # digest step, so there is no digest to name here.
  pub = ed25519_public_key(pub_raw)
  raise LicenseError, "bad signature" unless pub.verify(nil, signature, payload)

  # 3. Only now is the payload worth reading.
  lic = JSON.parse(payload)
  raise LicenseError, "unsupported license version" unless lic["v"] == 1

  lic
rescue JSON::ParserError
  raise LicenseError, "malformed payload"
rescue ArgumentError, OpenSSL::PKey::PKeyError
  raise LicenseError, "malformed public key"
end

# default_license_path is $XDG_CONFIG_HOME/shareware/licenses/<app>.key,
# falling back to ~/.config — where `omarket buy` writes the key.
def default_license_path
  base = ENV["XDG_CONFIG_HOME"]
  base = File.join(Dir.home, ".config") if base.nil? || base.empty?
  File.join(base, "shareware", "licenses", "#{APP_ID}.key")
end

# SHAREWARE_PUBLIC_KEY overrides the baked-in platform key — how you test
# against a local stack, or against examples/testdata.
def public_key
  env = ENV["SHAREWARE_PUBLIC_KEY"]
  env.nil? || env.empty? ? PLATFORM_PUBLIC_KEY : env
end

def nag(reason)
  puts "  [ ] unregistered — #{reason}"
  puts "      buy a key:  omarket buy #{APP_ID}"
  puts
  puts "  the deluxe feature is off until this app is registered."
end

def main(argv)
  puts "#{APP_NAME} 1.0"
  puts

  path = argv[0] || default_license_path
  unless File.readable?(path)
    nag("no license file at #{path}")
    return
  end

  lic = verify_license(File.read(path).strip, public_key)

  # Genuine — but is it ours? Skip this and any paid omarket license, for any
  # app, unlocks yours.
  if lic["app"] != APP_ID
    nag(%(key is for "#{lic["app"]}", not "#{APP_ID}"))
    return
  end

  puts "  [x] registered"
  puts "      license  #{lic["id"]}"
  puts "      kind     #{lic["kind"]}"
  puts "      issued   #{Time.at(lic["issued_at"]).utc.strftime("%Y-%m-%d")}"
  puts
  puts "  the deluxe feature is unlocked. Enjoy."
rescue LicenseError => e
  nag(e.message)
end

main(ARGV) if $PROGRAM_NAME == __FILE__
